{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bioNX: Automated Knowledge Graph Construction for PPI Networks\n",
    "\n",
    "Given \n",
    "\n",
    "This notebook contains a workflow that integrates data from several sources:\n",
    "* [bioGRID](https://thebiogrid.org/) - primary data source for PPIs\n",
    "* [HGNC](https://www.genenames.org/) - Gene nomenclature reference\n",
    "* [PubMed](https://pubmed.ncbi.nlm.nih.gov/) - Literature\n",
    "* [Uniprot](https://www.uniprot.org/) - Protein properties\n",
    "* [Entrez](https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html) - Gene properties\n",
    "* [GO](http://geneontology.org/) - Gene properties\n",
    "\n",
    "*Please note this project is under development.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path[0] = '../'\n",
    "from dotenv import load_dotenv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests\n",
    "from xml.etree.ElementTree import fromstring, ElementTree\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display options\n",
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set access key for BioGRID REST API\n",
    "load_dotenv()\n",
    "BIOGRID_ACCESS_KEY = os.getenv('BIOGRID_ACCESS_KEY')\n",
    "NEO4J_HOME = os.getenv('NEO4J_HOME')\n",
    "importDir = '/Users/gregory/Library/Application Support/Neo4j Desktop/Application/neo4jDatabases/database-d0b780b3-ce77-46cc-b7ed-bd0f78a46581/installation-3.5.14/import/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Building the Base Graph\n",
    "\n",
    "Interaction data is fetched from the BioGRID API for a single gene and its protein-protein interactions (PPIs). The PPI dataset forms the core foundation upon which the knowledge graph is built. It is augmented using data fetched from additional sources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loads data and formats columns\n",
    "def load_biogrid_data(gene_specifier, limit=10000):\n",
    "\n",
    "    if not isinstance(gene_specifier, list): \n",
    "        gene_specifier = gene_specifier\n",
    "    else:\n",
    "        gene_specifier = '|'.join(gene_specifier)\n",
    "    \n",
    "    url = f\"https://webservice.thebiogrid.org/interactions/?searchNames=true&geneList={gene_specifier}\" \\\n",
    "    \"&taxId=9606&includeInteractors=true&includeInteractorInteractions=true&includeHeader=true\" \\\n",
    "    f\"&accesskey={BIOGRID_ACCESS_KEY}&max={limit}\"\n",
    "\n",
    "    # Load data\n",
    "    data = pd.read_csv(url, sep='\\t', header=0)\n",
    "\n",
    "    # Remove leading hash character\n",
    "    data.rename(columns={\"#BioGRID Interaction ID\":\"BioGRID Interaction ID\"}, inplace=True)\n",
    "\n",
    "    # Replace pipe separators with commas\n",
    "    data = data.replace('\\|', ',', regex=True)\n",
    "\n",
    "    # Select str columns and replace '-' with np.nan\n",
    "    cols = ['Systematic Name Interactor A', \n",
    "          'Systematic Name Interactor B', \n",
    "          'Score', \n",
    "          'Modification', \n",
    "          'Phenotypes',\n",
    "          'Qualifications',\n",
    "          'Tags']\n",
    "\n",
    "    data[cols] = data[cols].applymap(lambda col: re.sub(r'^-$', str(np.NaN), col))\n",
    "       \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selects and transforms columns\n",
    "def preprocess_biogrid_data(data):\n",
    "    \n",
    "     # Convert Score column to float\n",
    "    data['Score'] = data['Score'].astype('float64')\n",
    "    \n",
    "    # Select columns of interest for graph\n",
    "    data = data[['BioGRID Interaction ID', 'Official Symbol Interactor A', 'Entrez Gene Interactor A', \n",
    "                       'Synonyms Interactor A', 'Organism Interactor A', 'Official Symbol Interactor B', \n",
    "                       'Entrez Gene Interactor B', 'Synonyms Interactor B', 'Organism Interactor B', \n",
    "                       'Author', 'Pubmed ID', 'Experimental System', 'Experimental System Type', 'Throughput']]\n",
    "\n",
    "    # Create Year column\n",
    "    data['Publication Year'] = data['Author'].str.split(' ').str[-1].str.strip('()')    \n",
    "    \n",
    "    # Remove Year from Author column\n",
    "    data['Author'] = data['Author'].str.split(' ').str[:2]\n",
    "    data['Author'] = data['Author'].apply(lambda x: ', '.join(x))\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "# Select gene or genes; create for loop and append dataframes for multiple genes\n",
    "gene = 'BRCA1'#'TP53'#'CYP1A2'\n",
    "genes = ['BRCA1', 'TP53']#['FECH', 'CYP1A2']\n",
    "\n",
    "# Fetch and clean data\n",
    "ppi_data = load_biogrid_data(gene, limit=10)\n",
    "ppi_data = preprocess_biogrid_data(ppi_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10, 15)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioGRID Interaction ID</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>Organism Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>Synonyms Interactor B</th>\n",
       "      <th>Organism Interactor B</th>\n",
       "      <th>Author</th>\n",
       "      <th>Pubmed ID</th>\n",
       "      <th>Experimental System</th>\n",
       "      <th>Experimental System Type</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Publication Year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1161</td>\n",
       "      <td>EP300</td>\n",
       "      <td>2033</td>\n",
       "      <td>KAT3B,RSTS2,p300</td>\n",
       "      <td>9606</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>7020</td>\n",
       "      <td>AP-2,AP-2alpha,AP2TF,BOFS,TFAP2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Braganca, J</td>\n",
       "      <td>12586840</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2368</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>672</td>\n",
       "      <td>BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>ATF1</td>\n",
       "      <td>466</td>\n",
       "      <td>EWS-ATF1,FUS/ATF-1,TREB36</td>\n",
       "      <td>9606</td>\n",
       "      <td>Houvras, Y</td>\n",
       "      <td>10945975</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2398</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>672</td>\n",
       "      <td>BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>4436</td>\n",
       "      <td>COCA1,FCC1,HNPCC,HNPCC1,LCFS2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Wang, Q</td>\n",
       "      <td>11498787</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2411</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>672</td>\n",
       "      <td>BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>BARD1</td>\n",
       "      <td>580</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Wu, LC</td>\n",
       "      <td>8944023</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>1996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2424</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>672</td>\n",
       "      <td>BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>MSH6</td>\n",
       "      <td>2956</td>\n",
       "      <td>GTBP,GTMBP,HNPCC5,HSAP,p160</td>\n",
       "      <td>9606</td>\n",
       "      <td>Wang, Q</td>\n",
       "      <td>11498787</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3315</td>\n",
       "      <td>MCM5</td>\n",
       "      <td>4174</td>\n",
       "      <td>CDC46,P1-CDC46</td>\n",
       "      <td>9606</td>\n",
       "      <td>MCM2</td>\n",
       "      <td>4171</td>\n",
       "      <td>BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19</td>\n",
       "      <td>9606</td>\n",
       "      <td>Kneissl, M</td>\n",
       "      <td>12614612</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3324</td>\n",
       "      <td>ORC2</td>\n",
       "      <td>4999</td>\n",
       "      <td>ORC2L</td>\n",
       "      <td>9606</td>\n",
       "      <td>MCM2</td>\n",
       "      <td>4171</td>\n",
       "      <td>BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19</td>\n",
       "      <td>9606</td>\n",
       "      <td>Kneissl, M</td>\n",
       "      <td>12614612</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3325</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>6118</td>\n",
       "      <td>REPA2,RP-A p32,RP-A p34,RPA32</td>\n",
       "      <td>9606</td>\n",
       "      <td>MCM2</td>\n",
       "      <td>4171</td>\n",
       "      <td>BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19</td>\n",
       "      <td>9606</td>\n",
       "      <td>Kneissl, M</td>\n",
       "      <td>12614612</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3326</td>\n",
       "      <td>DBF4</td>\n",
       "      <td>10926</td>\n",
       "      <td>ASK,CHIF,DBF4A,ZDBF1</td>\n",
       "      <td>9606</td>\n",
       "      <td>MCM2</td>\n",
       "      <td>4171</td>\n",
       "      <td>BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19</td>\n",
       "      <td>9606</td>\n",
       "      <td>Kneissl, M</td>\n",
       "      <td>12614612</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3329</td>\n",
       "      <td>AKAP8</td>\n",
       "      <td>10270</td>\n",
       "      <td>AKAP 95,AKAP-8,AKAP-95,AKAP95</td>\n",
       "      <td>9606</td>\n",
       "      <td>MCM2</td>\n",
       "      <td>4171</td>\n",
       "      <td>BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19</td>\n",
       "      <td>9606</td>\n",
       "      <td>Eide, T</td>\n",
       "      <td>12740381</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   BioGRID Interaction ID Official Symbol Interactor A  \\\n",
       "0                    1161                        EP300   \n",
       "1                    2368                        BRCA1   \n",
       "2                    2398                        BRCA1   \n",
       "3                    2411                        BRCA1   \n",
       "4                    2424                        BRCA1   \n",
       "5                    3315                         MCM5   \n",
       "6                    3324                         ORC2   \n",
       "7                    3325                         RPA2   \n",
       "8                    3326                         DBF4   \n",
       "9                    3329                        AKAP8   \n",
       "\n",
       "   Entrez Gene Interactor A  \\\n",
       "0                      2033   \n",
       "1                       672   \n",
       "2                       672   \n",
       "3                       672   \n",
       "4                       672   \n",
       "5                      4174   \n",
       "6                      4999   \n",
       "7                      6118   \n",
       "8                     10926   \n",
       "9                     10270   \n",
       "\n",
       "                               Synonyms Interactor A  Organism Interactor A  \\\n",
       "0                                   KAT3B,RSTS2,p300                   9606   \n",
       "1  BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...                   9606   \n",
       "2  BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...                   9606   \n",
       "3  BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...                   9606   \n",
       "4  BRCAI,BRCC1,BROVCA1,FANCS,IRIS,PNCA4,PPP1R53,P...                   9606   \n",
       "5                                     CDC46,P1-CDC46                   9606   \n",
       "6                                              ORC2L                   9606   \n",
       "7                      REPA2,RP-A p32,RP-A p34,RPA32                   9606   \n",
       "8                               ASK,CHIF,DBF4A,ZDBF1                   9606   \n",
       "9                      AKAP 95,AKAP-8,AKAP-95,AKAP95                   9606   \n",
       "\n",
       "  Official Symbol Interactor B  Entrez Gene Interactor B  \\\n",
       "0                       TFAP2A                      7020   \n",
       "1                         ATF1                       466   \n",
       "2                         MSH2                      4436   \n",
       "3                        BARD1                       580   \n",
       "4                         MSH6                      2956   \n",
       "5                         MCM2                      4171   \n",
       "6                         MCM2                      4171   \n",
       "7                         MCM2                      4171   \n",
       "8                         MCM2                      4171   \n",
       "9                         MCM2                      4171   \n",
       "\n",
       "                    Synonyms Interactor B  Organism Interactor B       Author  \\\n",
       "0         AP-2,AP-2alpha,AP2TF,BOFS,TFAP2                   9606  Braganca, J   \n",
       "1               EWS-ATF1,FUS/ATF-1,TREB36                   9606   Houvras, Y   \n",
       "2           COCA1,FCC1,HNPCC,HNPCC1,LCFS2                   9606      Wang, Q   \n",
       "3                                       -                   9606       Wu, LC   \n",
       "4             GTBP,GTMBP,HNPCC5,HSAP,p160                   9606      Wang, Q   \n",
       "5  BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19                   9606   Kneissl, M   \n",
       "6  BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19                   9606   Kneissl, M   \n",
       "7  BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19                   9606   Kneissl, M   \n",
       "8  BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19                   9606   Kneissl, M   \n",
       "9  BM28,CCNL1,CDCL1,D3S3194,MITOTIN,cdc19                   9606      Eide, T   \n",
       "\n",
       "   Pubmed ID Experimental System Experimental System Type       Throughput  \\\n",
       "0   12586840          Two-hybrid                 physical   Low Throughput   \n",
       "1   10945975          Two-hybrid                 physical   Low Throughput   \n",
       "2   11498787          Two-hybrid                 physical   Low Throughput   \n",
       "3    8944023          Two-hybrid                 physical   Low Throughput   \n",
       "4   11498787          Two-hybrid                 physical   Low Throughput   \n",
       "5   12614612          Two-hybrid                 physical  High Throughput   \n",
       "6   12614612          Two-hybrid                 physical  High Throughput   \n",
       "7   12614612          Two-hybrid                 physical  High Throughput   \n",
       "8   12614612          Two-hybrid                 physical  High Throughput   \n",
       "9   12740381          Two-hybrid                 physical   Low Throughput   \n",
       "\n",
       "  Publication Year  \n",
       "0             2003  \n",
       "1             2000  \n",
       "2             2001  \n",
       "3             1996  \n",
       "4             2001  \n",
       "5             2003  \n",
       "6             2003  \n",
       "7             2003  \n",
       "8             2003  \n",
       "9             2003  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save in project folder\n",
    "ppi_data.to_csv('../data/clean/biogrid_ppi_data.csv', index=False)\n",
    "\n",
    "# Save to Neo4j imports folder for LOAD CSV command\n",
    "ppi_data.to_csv(importDir + 'biogrid_ppi_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Author       Pubmed ID\n",
       "Braganca, J  12586840     1\n",
       "Eide, T      12740381     1\n",
       "Houvras, Y   10945975     1\n",
       "Kneissl, M   12614612     4\n",
       "Wang, Q      11498787     2\n",
       "Name: Pubmed ID, dtype: int64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_data.groupby('Author')['Pubmed ID'].value_counts()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrating Data from External APIs\n",
    "\n",
    "Fetching data from various APIs to build a table for genes:\n",
    "\n",
    "* HGNC\n",
    "* PubMed\n",
    "* Uniprot\n",
    "* Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create new columns for gene: Gene Description, NCBI url, Wikipedia url, full sequence url, chromosome, #bp\n",
    "### Create new columns for article: Article Title, Publication, Pubmed url\n",
    "# Create new column: subcellular location, condition, PTM/Processing, chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HGNC \n",
    "\n",
    "Parses XML from request. Documentation at:<br>\n",
    "https://www.genenames.org/help/rest/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_hgnc_data(ppi_data):\n",
    "    \n",
    "    def build_gene_dict(gene):\n",
    "\n",
    "        url = f\"http://rest.genenames.org/fetch/symbol/{gene}\";\n",
    "        r = requests.get(url)\n",
    "\n",
    "        tree = ElementTree(fromstring(r.text))\n",
    "        root = tree.getroot()\n",
    "\n",
    "        str_attribs = ['symbol', 'name', 'entrez_id', 'locus_type', 'location', 'ensembl_gene_id', \n",
    "                    'locus_group']\n",
    "\n",
    "        arr_attribs = ['pubmed_id', 'gene_group', 'uniprot_ids', 'omim_id']\n",
    "\n",
    "        gene_dict = dict()\n",
    "\n",
    "        # retrieve <str> attributes\n",
    "        for index, name in enumerate(str_attribs):\n",
    "            elements = root.findall(f\".//str[@name='{name}']\") # add .upper() method to gene symbol\n",
    "            try:\n",
    "                gene_dict[name] = [elements[0].text]\n",
    "            except Exception as e:\n",
    "                #print(f\"Error: {e}\") # Log errors \n",
    "                gene_dict[name] = ['NULL'] ### Keep symbol if no record is available; some genes are new research; use synonyms\n",
    "                continue\n",
    "\n",
    "        # retrieve <arr> attributes\n",
    "        for index, name in enumerate(arr_attribs):\n",
    "            elements = root.findall(f\".//arr[@name='{name}']/*\")\n",
    "            try:\n",
    "                gene_dict[name] = [elements[0].text]\n",
    "            except Exception as e:\n",
    "                #print(f\"Error: {e}\") # Log errors \n",
    "                gene_dict[name] = ['NULL']\n",
    "                continue\n",
    "\n",
    "        return gene_dict\n",
    "    \n",
    "    def build_gene_dataframe(ppi_data):\n",
    "\n",
    "        genes = list(ppi_data['Official Symbol Interactor A'].unique())\n",
    "\n",
    "        gene_dict_list = []\n",
    "\n",
    "        for i, gene in enumerate(genes):\n",
    "            gene_dict = build_gene_dict(gene)\n",
    "            gene_dict_list.append(gene_dict)\n",
    "            \n",
    "            # Throttle to less than 10 requests per second\n",
    "            if i != 0 and i % 10 == 0:\n",
    "                time.sleep(.5)\n",
    "\n",
    "        merge_gene_dict = {}\n",
    "\n",
    "        for key in gene_dict_list[0].keys():\n",
    "            merge_gene_dict[key] = [gene_dict_list[i][key][0] for i in range(len(gene_dict_list))]\n",
    "\n",
    "        return pd.DataFrame(merge_gene_dict)\n",
    "    \n",
    "    return build_gene_dataframe(ppi_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pubmed\n",
    "\n",
    "Example query: <br>\n",
    "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=9806834 - 'pubmed_id'\n",
    "\n",
    "--> can fetch comma delimited list for id to retrieve all in one request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_pubmed_data(data, column):\n",
    "    \n",
    "    def build_article_dict(pubmed_id, column):\n",
    "\n",
    "        url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id={pubmed_id}\";\n",
    "        r = requests.get(url)\n",
    "\n",
    "        tree = ElementTree(fromstring(r.text))\n",
    "        root = tree.getroot()\n",
    "\n",
    "        item_attribs = ['Title', 'FullJournalName', 'DOI']\n",
    "\n",
    "        article_dict = {column:[pubmed_id]}\n",
    "\n",
    "        # retrieve <Item> attributes\n",
    "        for index, Name in enumerate(item_attribs):\n",
    "            elements = root.findall(f\".//Item[@Name='{Name}']\") # add .upper() method to gene symbol\n",
    "            try:\n",
    "                article_dict[Name.lower()] = [elements[0].text]\n",
    "            except Exception as e:\n",
    "                print(f\"Error: {e}\") # Log errors \n",
    "                article_dict[Name.lower()] = ['NULL'] ### Keep symbol if no record is available; some genes are new research; use synonyms\n",
    "                continue\n",
    "\n",
    "        # Throttle to 10 requests per second\n",
    "        time.sleep(0.1)\n",
    "\n",
    "        article_dict['full_journal_name'] = article_dict.pop('fulljournalname')\n",
    "        \n",
    "        return article_dict\n",
    "    \n",
    "    def build_article_dataframe(data, column):\n",
    "\n",
    "        pubmed_id_list = list(data[column].unique())\n",
    "\n",
    "        article_dict_list = [build_article_dict(article, column) for article in pubmed_id_list]\n",
    "        #print(article_dict_list)\n",
    "        \n",
    "        merge_article_dict = {}\n",
    "\n",
    "        # Possible to use csv.writerows() for csv output\n",
    "        for item in article_dict_list:\n",
    "            #print(item)\n",
    "            for key in item.keys():\n",
    "                #print(key)\n",
    "                merge_article_dict[key] = [article_dict_list[i][key][0] for i in range(len(article_dict_list))]\n",
    "        \n",
    "        return pd.DataFrame(merge_article_dict)\n",
    "    \n",
    "    article_df = build_article_dataframe(data, column)\n",
    "    \n",
    "    return data.merge(article_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: list index out of range\n",
      "Error: list index out of range\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>location</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>gene_group</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>omim_id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>full_journal_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAP2K2</td>\n",
       "      <td>mitogen-activated protein kinase kinase 2</td>\n",
       "      <td>5605</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>ENSG00000126934</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>8388392</td>\n",
       "      <td>Mitogen-activated protein kinase kinases</td>\n",
       "      <td>P36507</td>\n",
       "      <td>601263</td>\n",
       "      <td>Cloning and characterization of two distinct h...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGRMC1</td>\n",
       "      <td>progesterone receptor membrane component 1</td>\n",
       "      <td>10857</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Xq24</td>\n",
       "      <td>ENSG00000101856</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>9705155</td>\n",
       "      <td>Membrane associated progesterone receptor family</td>\n",
       "      <td>O00264</td>\n",
       "      <td>300435</td>\n",
       "      <td>Cloning and tissue expression of two putative ...</td>\n",
       "      <td>10.1515/bchm.1998.379.7.907</td>\n",
       "      <td>Biological chemistry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PNKP</td>\n",
       "      <td>polynucleotide kinase 3'-phosphatase</td>\n",
       "      <td>11284</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>19q13.33</td>\n",
       "      <td>ENSG00000039650</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>10446192</td>\n",
       "      <td>HAD Asp-based non-protein phosphatases</td>\n",
       "      <td>Q96T60</td>\n",
       "      <td>605610</td>\n",
       "      <td>Molecular cloning of the human gene, PNKP, enc...</td>\n",
       "      <td>10.1074/jbc.274.34.24176</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CIAO1</td>\n",
       "      <td>cytosolic iron-sulfur assembly component 1</td>\n",
       "      <td>9391</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>2q11.2</td>\n",
       "      <td>ENSG00000144021</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>9556563</td>\n",
       "      <td>WD repeat domain containing</td>\n",
       "      <td>O76071</td>\n",
       "      <td>604333</td>\n",
       "      <td>Ciao 1 is a novel WD40 protein that interacts ...</td>\n",
       "      <td>10.1074/jbc.273.18.10880</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATR</td>\n",
       "      <td>ATR serine/threonine kinase</td>\n",
       "      <td>545</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>3q23</td>\n",
       "      <td>ENSG00000175054</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>8978690</td>\n",
       "      <td>Armadillo like helical domain containing</td>\n",
       "      <td>Q13535</td>\n",
       "      <td>601215</td>\n",
       "      <td>The Schizosaccharomyces pombe rad3 checkpoint ...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>The EMBO journal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol                                        name entrez_id  \\\n",
       "0  MAP2K2   mitogen-activated protein kinase kinase 2      5605   \n",
       "1  PGRMC1  progesterone receptor membrane component 1     10857   \n",
       "2    PNKP        polynucleotide kinase 3'-phosphatase     11284   \n",
       "3   CIAO1  cytosolic iron-sulfur assembly component 1      9391   \n",
       "4     ATR                 ATR serine/threonine kinase       545   \n",
       "\n",
       "                  locus_type  location  ensembl_gene_id          locus_group  \\\n",
       "0  gene with protein product   19p13.3  ENSG00000126934  protein-coding gene   \n",
       "1  gene with protein product      Xq24  ENSG00000101856  protein-coding gene   \n",
       "2  gene with protein product  19q13.33  ENSG00000039650  protein-coding gene   \n",
       "3  gene with protein product    2q11.2  ENSG00000144021  protein-coding gene   \n",
       "4  gene with protein product      3q23  ENSG00000175054  protein-coding gene   \n",
       "\n",
       "  pubmed_id                                        gene_group uniprot_ids  \\\n",
       "0   8388392          Mitogen-activated protein kinase kinases      P36507   \n",
       "1   9705155  Membrane associated progesterone receptor family      O00264   \n",
       "2  10446192            HAD Asp-based non-protein phosphatases      Q96T60   \n",
       "3   9556563                       WD repeat domain containing      O76071   \n",
       "4   8978690          Armadillo like helical domain containing      Q13535   \n",
       "\n",
       "  omim_id                                              title  \\\n",
       "0  601263  Cloning and characterization of two distinct h...   \n",
       "1  300435  Cloning and tissue expression of two putative ...   \n",
       "2  605610  Molecular cloning of the human gene, PNKP, enc...   \n",
       "3  604333  Ciao 1 is a novel WD40 protein that interacts ...   \n",
       "4  601215  The Schizosaccharomyces pombe rad3 checkpoint ...   \n",
       "\n",
       "                           doi                    full_journal_name  \n",
       "0                         NULL  The Journal of biological chemistry  \n",
       "1  10.1515/bchm.1998.379.7.907                 Biological chemistry  \n",
       "2     10.1074/jbc.274.34.24176  The Journal of biological chemistry  \n",
       "3     10.1074/jbc.273.18.10880  The Journal of biological chemistry  \n",
       "4                         NULL                     The EMBO journal  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fetch_pubmed_data(gene_data, 'pubmed_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "article_dict_list = [{'pubmed_id': ['8388392'], 'title': ['Cloning and characterization of two distinct human extracellular signal-regulated kinase activator kinases, MEK1 and MEK2.'], 'DOI': ['NULL'], 'full_journal_name': ['The Journal of biological chemistry']}, {'pubmed_id': ['9705155'], 'title': ['Cloning and tissue expression of two putative steroid membrane receptors.'], 'doi': ['10.1515/bchm.1998.379.7.907'], 'full_journal_name': ['Biological chemistry']}, {'pubmed_id': ['10446192'], 'title': [\"Molecular cloning of the human gene, PNKP, encoding a polynucleotide kinase 3'-phosphatase and evidence for its role in repair of DNA strand breaks caused by oxidative damage.\"], 'doi': ['10.1074/jbc.274.34.24176'], 'full_journal_name': ['The Journal of biological chemistry']}, {'pubmed_id': ['9556563'], 'title': ['Ciao 1 is a novel WD40 protein that interacts with the tumor suppressor protein WT1.'], 'doi': ['10.1074/jbc.273.18.10880'], 'full_journal_name': ['The Journal of biological chemistry']}, {'pubmed_id': ['8978690'], 'title': ['The Schizosaccharomyces pombe rad3 checkpoint gene.'], 'DOI': ['NULL'], 'full_journal_name': ['The EMBO journal']}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pubmed_id': ['8388392'],\n",
       "  'title': ['Cloning and characterization of two distinct human extracellular signal-regulated kinase activator kinases, MEK1 and MEK2.'],\n",
       "  'DOI': ['NULL'],\n",
       "  'full_journal_name': ['The Journal of biological chemistry']},\n",
       " {'pubmed_id': ['9705155'],\n",
       "  'title': ['Cloning and tissue expression of two putative steroid membrane receptors.'],\n",
       "  'doi': ['10.1515/bchm.1998.379.7.907'],\n",
       "  'full_journal_name': ['Biological chemistry']},\n",
       " {'pubmed_id': ['10446192'],\n",
       "  'title': [\"Molecular cloning of the human gene, PNKP, encoding a polynucleotide kinase 3'-phosphatase and evidence for its role in repair of DNA strand breaks caused by oxidative damage.\"],\n",
       "  'doi': ['10.1074/jbc.274.34.24176'],\n",
       "  'full_journal_name': ['The Journal of biological chemistry']},\n",
       " {'pubmed_id': ['9556563'],\n",
       "  'title': ['Ciao 1 is a novel WD40 protein that interacts with the tumor suppressor protein WT1.'],\n",
       "  'doi': ['10.1074/jbc.273.18.10880'],\n",
       "  'full_journal_name': ['The Journal of biological chemistry']},\n",
       " {'pubmed_id': ['8978690'],\n",
       "  'title': ['The Schizosaccharomyces pombe rad3 checkpoint gene.'],\n",
       "  'DOI': ['NULL'],\n",
       "  'full_journal_name': ['The EMBO journal']}]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_dict_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, item in enumerate(article_dict_list):\n",
    "    #print(item.keys())\n",
    "    #article_dict_list.pop(idx)\n",
    "    for key in item:\n",
    "        item[key.lower()] = item.pop(key)\n",
    "        #print(item)\n",
    "    article_dict_list.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "article_dict_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean Final Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def final_process_data(data):\n",
    "    \n",
    "    def rename_columns(data):\n",
    "        data.columns = data.columns.str.lower().str.replace(' ', '_')\n",
    "        \n",
    "        return data\n",
    "\n",
    "    # Data cleaning pipeline\n",
    "    data = rename_columns(data)\n",
    "    \n",
    "    # Extract chromosome\n",
    "    \n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "# Select gene; create for loop and append dataframes for multiple genes\n",
    "gene = 'CYP1A2'#'CYP1A2'#'TP53'#'CYP1A2' # or\n",
    "#genes = ['FECH', 'CYP1A2']\n",
    "\n",
    "# Fetch and clean interaction data from BioGRID\n",
    "ppi_data = load_biogrid_data(gene, limit=20)\n",
    "ppi_data = preprocess_biogrid_data(ppi_data)\n",
    "\n",
    "# Fetch HGNC data for information on genes; time limit on requests s \n",
    "gene_data = fetch_hgnc_data(ppi_data)\n",
    "\n",
    "# Fetch PubMed data\n",
    "ppi_data = fetch_pubmed_data(ppi_data, 'Pubmed ID')\n",
    "gene_data = fetch_pubmed_data(gene_data, 'pubmed_id')\n",
    "\n",
    "# Final cleaning step\n",
    "ppi_data = final_process_data(ppi_data)\n",
    "gene_data = final_process_data(gene_data)\n",
    "\n",
    "# Save in project folder\n",
    "ppi_data.to_csv('../data/clean/biogrid_ppi_data.csv', index=False)\n",
    "gene_data.to_csv('../data/clean/gene_data.csv', index=False)\n",
    "\n",
    "# Save to Neo4j imports folder for LOAD CSV command\n",
    "ppi_data.to_csv(importDir + 'biogrid_ppi_data.csv', index=False)\n",
    "gene_data.to_csv(importDir + 'gene_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test for all genes present in genes data\n",
    "interaction_set = set(interaction_data['Official Symbol Interactor A'].unique())\n",
    "gene_set = set(gene_data['symbol'].unique())\n",
    "\n",
    "assert interaction_set == gene_set, \\\n",
    "    f\"There are genes missing in the data:\\ninteraction data:{interaction_set - gene_set}\\ngene data: {gene_set - interaction_set}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>location</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>gene_group</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>omim_id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>full_journal_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FECH</td>\n",
       "      <td>ferrochelatase</td>\n",
       "      <td>2235</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>18q21.31</td>\n",
       "      <td>ENSG00000066926</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>1838349</td>\n",
       "      <td>NULL</td>\n",
       "      <td>P22830</td>\n",
       "      <td>612386</td>\n",
       "      <td>Assignment of the gene for cyclic AMP-response...</td>\n",
       "      <td>10.1016/0888-7543(91)90047-i</td>\n",
       "      <td>Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>POR</td>\n",
       "      <td>cytochrome p450 oxidoreductase</td>\n",
       "      <td>5447</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>7q11.23</td>\n",
       "      <td>ENSG00000127948</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>2516426</td>\n",
       "      <td>MicroRNA protein coding host genes</td>\n",
       "      <td>P16435</td>\n",
       "      <td>124015</td>\n",
       "      <td>Isolation of a human cytochrome P-450 reductas...</td>\n",
       "      <td>10.1111/j.1469-1809.1989.tb01798.x</td>\n",
       "      <td>Annals of human genetics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGRMC1</td>\n",
       "      <td>progesterone receptor membrane component 1</td>\n",
       "      <td>10857</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Xq24</td>\n",
       "      <td>ENSG00000101856</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>9705155</td>\n",
       "      <td>Membrane associated progesterone receptor family</td>\n",
       "      <td>O00264</td>\n",
       "      <td>300435</td>\n",
       "      <td>Cloning and tissue expression of two putative ...</td>\n",
       "      <td>10.1515/bchm.1998.379.7.907</td>\n",
       "      <td>Biological chemistry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>cytochrome P450 family 1 subfamily A member 2</td>\n",
       "      <td>1544</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>15q24.1</td>\n",
       "      <td>ENSG00000140505</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>15128046</td>\n",
       "      <td>Cytochrome P450 family 1</td>\n",
       "      <td>P05177</td>\n",
       "      <td>124060</td>\n",
       "      <td>Comparison of cytochrome P450 (CYP) genes from...</td>\n",
       "      <td>10.1097/00008571-200401000-00001</td>\n",
       "      <td>Pharmacogenetics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CYB5A</td>\n",
       "      <td>cytochrome b5 type A</td>\n",
       "      <td>1528</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>18q22.3</td>\n",
       "      <td>ENSG00000166347</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>1840560</td>\n",
       "      <td>NULL</td>\n",
       "      <td>P00167</td>\n",
       "      <td>613218</td>\n",
       "      <td>Chromosomal localization of a cytochrome b5 ge...</td>\n",
       "      <td>10.1016/0888-7543(91)90136-3</td>\n",
       "      <td>Genomics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol                                           name entrez_id  \\\n",
       "0    FECH                                 ferrochelatase      2235   \n",
       "1     POR                 cytochrome p450 oxidoreductase      5447   \n",
       "2  PGRMC1     progesterone receptor membrane component 1     10857   \n",
       "3  CYP1A2  cytochrome P450 family 1 subfamily A member 2      1544   \n",
       "4   CYB5A                           cytochrome b5 type A      1528   \n",
       "\n",
       "                  locus_type  location  ensembl_gene_id          locus_group  \\\n",
       "0  gene with protein product  18q21.31  ENSG00000066926  protein-coding gene   \n",
       "1  gene with protein product   7q11.23  ENSG00000127948  protein-coding gene   \n",
       "2  gene with protein product      Xq24  ENSG00000101856  protein-coding gene   \n",
       "3  gene with protein product   15q24.1  ENSG00000140505  protein-coding gene   \n",
       "4  gene with protein product   18q22.3  ENSG00000166347  protein-coding gene   \n",
       "\n",
       "  pubmed_id                                        gene_group uniprot_ids  \\\n",
       "0   1838349                                              NULL      P22830   \n",
       "1   2516426                MicroRNA protein coding host genes      P16435   \n",
       "2   9705155  Membrane associated progesterone receptor family      O00264   \n",
       "3  15128046                          Cytochrome P450 family 1      P05177   \n",
       "4   1840560                                              NULL      P00167   \n",
       "\n",
       "  omim_id                                              title  \\\n",
       "0  612386  Assignment of the gene for cyclic AMP-response...   \n",
       "1  124015  Isolation of a human cytochrome P-450 reductas...   \n",
       "2  300435  Cloning and tissue expression of two putative ...   \n",
       "3  124060  Comparison of cytochrome P450 (CYP) genes from...   \n",
       "4  613218  Chromosomal localization of a cytochrome b5 ge...   \n",
       "\n",
       "                                  doi         full_journal_name  \n",
       "0        10.1016/0888-7543(91)90047-i                  Genomics  \n",
       "1  10.1111/j.1469-1809.1989.tb01798.x  Annals of human genetics  \n",
       "2         10.1515/bchm.1998.379.7.907      Biological chemistry  \n",
       "3    10.1097/00008571-200401000-00001          Pharmacogenetics  \n",
       "4        10.1016/0888-7543(91)90136-3                  Genomics  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
