{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bioNX: Automated Knowledge Graph Construction for PPI Networks\n",
    "\n",
    "Given \n",
    "\n",
    "This notebook contains a workflow that integrates data from several sources:\n",
    "* [bioGRID](https://thebiogrid.org/) - primary data source for PPIs\n",
    "* [HGNC](https://www.genenames.org/) - Gene nomenclature reference\n",
    "* [PubMed](https://pubmed.ncbi.nlm.nih.gov/) - Literature\n",
    "* [Uniprot](https://www.uniprot.org/) - Protein properties (*pending implementation*)\n",
    "* [Entrez](https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html) - Gene properties (*pending implementation*)\n",
    "* [GO](http://geneontology.org/) - Gene properties (*pending implementation*)\n",
    "\n",
    "*Please note this project is under development.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path[0] = '../'\n",
    "from dotenv import load_dotenv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests\n",
    "from xml.etree.ElementTree import fromstring, ElementTree\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display options\n",
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set access key for BioGRID REST API\n",
    "load_dotenv()\n",
    "BIOGRID_ACCESS_KEY = os.getenv('BIOGRID_ACCESS_KEY')\n",
    "NEO4J_HOME = os.getenv('NEO4J_HOME')\n",
    "importDir = NEO4J_HOME.replace('\\\\', '') + 'import/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Building the Base Graph\n",
    "\n",
    "Interaction data is fetched from the BioGRID API for a single gene and its protein-protein interactions (PPIs). The PPI dataset forms the core foundation upon which the knowledge graph is built. It is augmented using data fetched from additional sources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loads data and formats columns\n",
    "def load_biogrid_data(gene_specifier, limit=10000):\n",
    "\n",
    "    if not isinstance(gene_specifier, list): \n",
    "        gene_specifier = gene_specifier\n",
    "    else:\n",
    "        gene_specifier = '|'.join(gene_specifier)\n",
    "    \n",
    "    url = f\"https://webservice.thebiogrid.org/interactions/?searchNames=true&geneList={gene_specifier}\" \\\n",
    "    \"&taxId=9606&includeInteractors=true&includeInteractorInteractions=true&includeHeader=true\" \\\n",
    "    f\"&accesskey={BIOGRID_ACCESS_KEY}&max={limit}\"\n",
    "\n",
    "    # Load data\n",
    "    data = pd.read_csv(url, sep='\\t', header=0)\n",
    "\n",
    "    # Remove leading hash character\n",
    "    data.rename(columns={\"#BioGRID Interaction ID\":\"BioGRID Interaction ID\"}, inplace=True)\n",
    "\n",
    "    # Replace pipe separators with commas\n",
    "    data = data.replace('\\|', ',', regex=True)\n",
    "\n",
    "    # Select str columns and replace '-' with np.nan\n",
    "    cols = ['Systematic Name Interactor A', \n",
    "          'Systematic Name Interactor B', \n",
    "          'Score', \n",
    "          'Modification', \n",
    "          'Phenotypes',\n",
    "          'Qualifications',\n",
    "          'Tags']\n",
    "\n",
    "    data[cols] = data[cols].applymap(lambda col: re.sub(r'^-$', str(np.NaN), col))\n",
    "       \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selects and transforms columns\n",
    "def preprocess_biogrid_data(data):\n",
    "    \n",
    "     # Convert Score column to float\n",
    "    data['Score'] = data['Score'].astype('float64')\n",
    "    \n",
    "    # Select columns of interest for graph\n",
    "    data = data[['BioGRID Interaction ID', 'Official Symbol Interactor A', 'Entrez Gene Interactor A', \n",
    "                       'Synonyms Interactor A', 'Organism Interactor A', 'Official Symbol Interactor B', \n",
    "                       'Entrez Gene Interactor B', 'Synonyms Interactor B', 'Organism Interactor B', \n",
    "                       'Author', 'Pubmed ID', 'Experimental System', 'Experimental System Type', 'Throughput']]\n",
    "\n",
    "    # Create Year column\n",
    "    data['Publication Year'] = data['Author'].str.split(' ').str[-1].str.strip('()')    \n",
    "    \n",
    "    # Remove Year from Author column\n",
    "    data['Author'] = data['Author'].str.split(' ').str[:2]\n",
    "    data['Author'] = data['Author'].apply(lambda x: ', '.join(x))\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrating Data from External APIs\n",
    "\n",
    "Fetching data from various APIs to build a table for genes:\n",
    "\n",
    "* HGNC\n",
    "* PubMed\n",
    "* Uniprot\n",
    "* Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create new columns for gene: Gene Description, NCBI url, Wikipedia url, full sequence url, chromosome, base pairs\n",
    "### Create new columns for article: Article Title, Publication, Pubmed url\n",
    "### Create new column: subcellular location, condition, PTM/Processing, chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HGNC\n",
    "\n",
    "Parses XML from request. Documentation at:<br>\n",
    "https://www.genenames.org/help/rest/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_hgnc_data(ppi_data):\n",
    "    \n",
    "    def build_gene_dict(gene):\n",
    "\n",
    "        url = f\"http://rest.genenames.org/fetch/symbol/{gene}\";\n",
    "        r = requests.get(url)\n",
    "\n",
    "        tree = ElementTree(fromstring(r.text))\n",
    "        root = tree.getroot()\n",
    "\n",
    "        str_attribs = ['symbol', 'name', 'entrez_id', 'locus_type', 'location', 'ensembl_gene_id', \n",
    "                    'locus_group']\n",
    "\n",
    "        arr_attribs = ['pubmed_id', 'gene_group', 'uniprot_ids', 'omim_id']\n",
    "\n",
    "        gene_dict = dict()\n",
    "\n",
    "        # retrieve <str> attributes\n",
    "        for index, name in enumerate(str_attribs):\n",
    "            elements = root.findall(f\".//str[@name='{name}']\") # add .upper() method to gene symbol\n",
    "            try:\n",
    "                gene_dict[name] = [elements[0].text]\n",
    "            except Exception as e:\n",
    "                #print(f\"Error: {e}\") # Log errors \n",
    "                gene_dict[name] = ['NULL'] ### Keep symbol if no record is available; some genes are new research; use synonyms\n",
    "                continue\n",
    "\n",
    "        # retrieve <arr> attributes\n",
    "        for index, name in enumerate(arr_attribs):\n",
    "            elements = root.findall(f\".//arr[@name='{name}']/*\")\n",
    "            try:\n",
    "                gene_dict[name] = [elements[0].text]\n",
    "            except Exception as e:\n",
    "                #print(f\"Error: {e}\") # Log errors \n",
    "                gene_dict[name] = ['NULL']\n",
    "                continue\n",
    "\n",
    "        return gene_dict\n",
    "    \n",
    "    def build_gene_dataframe(ppi_data):\n",
    "\n",
    "        genes = list(ppi_data['Official Symbol Interactor A'].unique())\n",
    "\n",
    "        gene_dict_list = []\n",
    "\n",
    "        for i, gene in enumerate(genes):\n",
    "            gene_dict = build_gene_dict(gene)\n",
    "            gene_dict_list.append(gene_dict)\n",
    "            \n",
    "            # Throttle to less than 10 requests per second\n",
    "            if i != 0 and i % 10 == 0:\n",
    "                time.sleep(.5)\n",
    "\n",
    "        merge_gene_dict = {}\n",
    "\n",
    "        for key in gene_dict_list[0].keys():\n",
    "            merge_gene_dict[key] = [gene_dict_list[i][key][0] for i in range(len(gene_dict_list))]\n",
    "\n",
    "        return pd.DataFrame(merge_gene_dict)\n",
    "    \n",
    "    return build_gene_dataframe(ppi_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pubmed\n",
    "\n",
    "Example query: <br>\n",
    "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=9806834 - 'pubmed_id'\n",
    "\n",
    "--> can fetch comma delimited list for id to retrieve all in one request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_pubmed_data(data, column):\n",
    "    \n",
    "    def build_article_dict(pubmed_id, column):\n",
    "\n",
    "        url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id={pubmed_id}\";\n",
    "        r = requests.get(url)\n",
    "\n",
    "        tree = ElementTree(fromstring(r.text))\n",
    "        root = tree.getroot()\n",
    "\n",
    "        item_attribs = ['Title', 'FullJournalName', 'DOI']\n",
    "\n",
    "        article_dict = {column:[pubmed_id]}\n",
    "\n",
    "        # retrieve <Item> attributes\n",
    "        for index, Name in enumerate(item_attribs):\n",
    "            elements = root.findall(f\".//Item[@Name='{Name}']\") # add .upper() method to gene symbol\n",
    "            try:\n",
    "                article_dict[Name.lower()] = [elements[0].text]\n",
    "            except Exception as e:\n",
    "                print(f\"Error: {e}\") # Log errors \n",
    "                article_dict[Name.lower()] = ['NULL'] ### Keep symbol if no record is available; some genes are new research; use synonyms\n",
    "                continue\n",
    "\n",
    "        # Throttle to 10 requests per second\n",
    "        time.sleep(0.1)\n",
    "\n",
    "        article_dict['full_journal_name'] = article_dict.pop('fulljournalname')\n",
    "        \n",
    "        return article_dict\n",
    "    \n",
    "    def build_article_dataframe(data, column):\n",
    "\n",
    "        pubmed_id_list = list(data[column].unique())\n",
    "\n",
    "        article_dict_list = [build_article_dict(article, column) for article in pubmed_id_list]\n",
    "        #print(article_dict_list)\n",
    "        \n",
    "        merge_article_dict = {}\n",
    "\n",
    "        # Possible to use csv.writerows() for csv output\n",
    "        for item in article_dict_list:\n",
    "            #print(item)\n",
    "            for key in item.keys():\n",
    "                #print(key)\n",
    "                merge_article_dict[key] = [article_dict_list[i][key][0] for i in range(len(article_dict_list))]\n",
    "        \n",
    "        return pd.DataFrame(merge_article_dict)\n",
    "    \n",
    "    article_df = build_article_dataframe(data, column)\n",
    "    \n",
    "    return data.merge(article_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Final Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def final_process_data(data):\n",
    "    \n",
    "    def rename_columns(data):\n",
    "        data.columns = data.columns.str.lower().str.replace(' ', '_')\n",
    "        \n",
    "        return data\n",
    "\n",
    "    # Data cleaning pipeline\n",
    "    data = rename_columns(data)\n",
    "    \n",
    "    # Extract chromosome\n",
    "    \n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate the Knowledge Graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'MTHFR'\n",
    "limit = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run the Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch and clean interaction data from BioGRID\n",
    "ppi_data = load_biogrid_data(gene, limit=limit)\n",
    "ppi_data = preprocess_biogrid_data(ppi_data)\n",
    "\n",
    "# Fetch HGNC data for information on genes; time limit on requests s \n",
    "gene_data = fetch_hgnc_data(ppi_data)\n",
    "\n",
    "# Fetch PubMed data\n",
    "ppi_data = fetch_pubmed_data(ppi_data, 'Pubmed ID')\n",
    "gene_data = fetch_pubmed_data(gene_data, 'pubmed_id')\n",
    "\n",
    "# Final cleaning step\n",
    "ppi_data = final_process_data(ppi_data)\n",
    "gene_data = final_process_data(gene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biogrid_interaction_id</th>\n",
       "      <th>official_symbol_interactor_a</th>\n",
       "      <th>entrez_gene_interactor_a</th>\n",
       "      <th>synonyms_interactor_a</th>\n",
       "      <th>organism_interactor_a</th>\n",
       "      <th>official_symbol_interactor_b</th>\n",
       "      <th>entrez_gene_interactor_b</th>\n",
       "      <th>synonyms_interactor_b</th>\n",
       "      <th>organism_interactor_b</th>\n",
       "      <th>author</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>experimental_system</th>\n",
       "      <th>experimental_system_type</th>\n",
       "      <th>throughput</th>\n",
       "      <th>publication_year</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>full_journal_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>287055</td>\n",
       "      <td>RASSF1</td>\n",
       "      <td>11186</td>\n",
       "      <td>123F2,NORE2A,RASSF1A,RDA32,REH3P21</td>\n",
       "      <td>9606</td>\n",
       "      <td>RASSF1</td>\n",
       "      <td>11186</td>\n",
       "      <td>123F2,NORE2A,RASSF1A,RDA32,REH3P21</td>\n",
       "      <td>9606</td>\n",
       "      <td>Ortiz-Vega, S</td>\n",
       "      <td>11857081</td>\n",
       "      <td>Affinity Capture-Western</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2002</td>\n",
       "      <td>The putative tumor suppressor RASSF1A homodime...</td>\n",
       "      <td>10.1038/sj.onc.1205192</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>719692</td>\n",
       "      <td>PRPF31</td>\n",
       "      <td>26121</td>\n",
       "      <td>NY-BR-99,PRP31,RP11,SNRNP61</td>\n",
       "      <td>9606</td>\n",
       "      <td>PRPF31</td>\n",
       "      <td>26121</td>\n",
       "      <td>NY-BR-99,PRP31,RP11,SNRNP61</td>\n",
       "      <td>9606</td>\n",
       "      <td>Liu, S</td>\n",
       "      <td>16723661</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2006</td>\n",
       "      <td>The network of protein-protein interactions wi...</td>\n",
       "      <td>10.1261/rna.55406</td>\n",
       "      <td>RNA (New York, N.Y.)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>815851</td>\n",
       "      <td>CDCA8</td>\n",
       "      <td>55143</td>\n",
       "      <td>BOR,BOREALIN,DasraB,MESRGP</td>\n",
       "      <td>9606</td>\n",
       "      <td>CDCA8</td>\n",
       "      <td>55143</td>\n",
       "      <td>BOR,BOREALIN,DasraB,MESRGP</td>\n",
       "      <td>9606</td>\n",
       "      <td>Klein, UR</td>\n",
       "      <td>16571674</td>\n",
       "      <td>Affinity Capture-Western</td>\n",
       "      <td>physical</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>2006</td>\n",
       "      <td>Centromere targeting of the chromosomal passen...</td>\n",
       "      <td>10.1091/mbc.e05-12-1133</td>\n",
       "      <td>Molecular biology of the cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>905406</td>\n",
       "      <td>LSM8</td>\n",
       "      <td>51691</td>\n",
       "      <td>NAA38</td>\n",
       "      <td>9606</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>4524</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Lehner, B</td>\n",
       "      <td>15231747</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2004</td>\n",
       "      <td>A protein interaction framework for human mRNA...</td>\n",
       "      <td>10.1101/gr.2122004</td>\n",
       "      <td>Genome research</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1038550</td>\n",
       "      <td>GNAZ</td>\n",
       "      <td>2781</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>GNAZ</td>\n",
       "      <td>2781</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Rolland, T</td>\n",
       "      <td>25416956</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2014</td>\n",
       "      <td>A proteome-scale map of the human interactome ...</td>\n",
       "      <td>10.1016/j.cell.2014.10.050</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1179975</td>\n",
       "      <td>CDCA8</td>\n",
       "      <td>55143</td>\n",
       "      <td>BOR,BOREALIN,DasraB,MESRGP</td>\n",
       "      <td>9606</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>4524</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Huttlin, EL</td>\n",
       "      <td>26186194</td>\n",
       "      <td>Affinity Capture-MS</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2015</td>\n",
       "      <td>The BioPlex Network: A Systematic Exploration ...</td>\n",
       "      <td>10.1016/j.cell.2015.06.043</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1186580</td>\n",
       "      <td>GNAI1</td>\n",
       "      <td>2770</td>\n",
       "      <td>Gi</td>\n",
       "      <td>9606</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>4524</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Huttlin, EL</td>\n",
       "      <td>26186194</td>\n",
       "      <td>Affinity Capture-MS</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2015</td>\n",
       "      <td>The BioPlex Network: A Systematic Exploration ...</td>\n",
       "      <td>10.1016/j.cell.2015.06.043</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1186621</td>\n",
       "      <td>GNAZ</td>\n",
       "      <td>2781</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>4524</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Huttlin, EL</td>\n",
       "      <td>26186194</td>\n",
       "      <td>Affinity Capture-MS</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2015</td>\n",
       "      <td>The BioPlex Network: A Systematic Exploration ...</td>\n",
       "      <td>10.1016/j.cell.2015.06.043</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1188394</td>\n",
       "      <td>RASSF1</td>\n",
       "      <td>11186</td>\n",
       "      <td>123F2,NORE2A,RASSF1A,RDA32,REH3P21</td>\n",
       "      <td>9606</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>4524</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Huttlin, EL</td>\n",
       "      <td>26186194</td>\n",
       "      <td>Affinity Capture-MS</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2015</td>\n",
       "      <td>The BioPlex Network: A Systematic Exploration ...</td>\n",
       "      <td>10.1016/j.cell.2015.06.043</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1191762</td>\n",
       "      <td>SMPD2</td>\n",
       "      <td>6610</td>\n",
       "      <td>ISC1,NSMASE,NSMASE1</td>\n",
       "      <td>9606</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>4524</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>Huttlin, EL</td>\n",
       "      <td>26186194</td>\n",
       "      <td>Affinity Capture-MS</td>\n",
       "      <td>physical</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>2015</td>\n",
       "      <td>The BioPlex Network: A Systematic Exploration ...</td>\n",
       "      <td>10.1016/j.cell.2015.06.043</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biogrid_interaction_id official_symbol_interactor_a  \\\n",
       "0                  287055                       RASSF1   \n",
       "1                  719692                       PRPF31   \n",
       "2                  815851                        CDCA8   \n",
       "3                  905406                         LSM8   \n",
       "4                 1038550                         GNAZ   \n",
       "5                 1179975                        CDCA8   \n",
       "6                 1186580                        GNAI1   \n",
       "7                 1186621                         GNAZ   \n",
       "8                 1188394                       RASSF1   \n",
       "9                 1191762                        SMPD2   \n",
       "\n",
       "   entrez_gene_interactor_a               synonyms_interactor_a  \\\n",
       "0                     11186  123F2,NORE2A,RASSF1A,RDA32,REH3P21   \n",
       "1                     26121         NY-BR-99,PRP31,RP11,SNRNP61   \n",
       "2                     55143          BOR,BOREALIN,DasraB,MESRGP   \n",
       "3                     51691                               NAA38   \n",
       "4                      2781                                   -   \n",
       "5                     55143          BOR,BOREALIN,DasraB,MESRGP   \n",
       "6                      2770                                  Gi   \n",
       "7                      2781                                   -   \n",
       "8                     11186  123F2,NORE2A,RASSF1A,RDA32,REH3P21   \n",
       "9                      6610                 ISC1,NSMASE,NSMASE1   \n",
       "\n",
       "   organism_interactor_a official_symbol_interactor_b  \\\n",
       "0                   9606                       RASSF1   \n",
       "1                   9606                       PRPF31   \n",
       "2                   9606                        CDCA8   \n",
       "3                   9606                        MTHFR   \n",
       "4                   9606                         GNAZ   \n",
       "5                   9606                        MTHFR   \n",
       "6                   9606                        MTHFR   \n",
       "7                   9606                        MTHFR   \n",
       "8                   9606                        MTHFR   \n",
       "9                   9606                        MTHFR   \n",
       "\n",
       "   entrez_gene_interactor_b               synonyms_interactor_b  \\\n",
       "0                     11186  123F2,NORE2A,RASSF1A,RDA32,REH3P21   \n",
       "1                     26121         NY-BR-99,PRP31,RP11,SNRNP61   \n",
       "2                     55143          BOR,BOREALIN,DasraB,MESRGP   \n",
       "3                      4524                                   -   \n",
       "4                      2781                                   -   \n",
       "5                      4524                                   -   \n",
       "6                      4524                                   -   \n",
       "7                      4524                                   -   \n",
       "8                      4524                                   -   \n",
       "9                      4524                                   -   \n",
       "\n",
       "   organism_interactor_b         author  pubmed_id       experimental_system  \\\n",
       "0                   9606  Ortiz-Vega, S   11857081  Affinity Capture-Western   \n",
       "1                   9606         Liu, S   16723661                Two-hybrid   \n",
       "2                   9606      Klein, UR   16571674  Affinity Capture-Western   \n",
       "3                   9606      Lehner, B   15231747                Two-hybrid   \n",
       "4                   9606     Rolland, T   25416956                Two-hybrid   \n",
       "5                   9606    Huttlin, EL   26186194       Affinity Capture-MS   \n",
       "6                   9606    Huttlin, EL   26186194       Affinity Capture-MS   \n",
       "7                   9606    Huttlin, EL   26186194       Affinity Capture-MS   \n",
       "8                   9606    Huttlin, EL   26186194       Affinity Capture-MS   \n",
       "9                   9606    Huttlin, EL   26186194       Affinity Capture-MS   \n",
       "\n",
       "  experimental_system_type       throughput publication_year  \\\n",
       "0                 physical   Low Throughput             2002   \n",
       "1                 physical   Low Throughput             2006   \n",
       "2                 physical   Low Throughput             2006   \n",
       "3                 physical  High Throughput             2004   \n",
       "4                 physical  High Throughput             2014   \n",
       "5                 physical  High Throughput             2015   \n",
       "6                 physical  High Throughput             2015   \n",
       "7                 physical  High Throughput             2015   \n",
       "8                 physical  High Throughput             2015   \n",
       "9                 physical  High Throughput             2015   \n",
       "\n",
       "                                               title  \\\n",
       "0  The putative tumor suppressor RASSF1A homodime...   \n",
       "1  The network of protein-protein interactions wi...   \n",
       "2  Centromere targeting of the chromosomal passen...   \n",
       "3  A protein interaction framework for human mRNA...   \n",
       "4  A proteome-scale map of the human interactome ...   \n",
       "5  The BioPlex Network: A Systematic Exploration ...   \n",
       "6  The BioPlex Network: A Systematic Exploration ...   \n",
       "7  The BioPlex Network: A Systematic Exploration ...   \n",
       "8  The BioPlex Network: A Systematic Exploration ...   \n",
       "9  The BioPlex Network: A Systematic Exploration ...   \n",
       "\n",
       "                          doi              full_journal_name  \n",
       "0      10.1038/sj.onc.1205192                       Oncogene  \n",
       "1           10.1261/rna.55406           RNA (New York, N.Y.)  \n",
       "2     10.1091/mbc.e05-12-1133  Molecular biology of the cell  \n",
       "3          10.1101/gr.2122004                Genome research  \n",
       "4  10.1016/j.cell.2014.10.050                           Cell  \n",
       "5  10.1016/j.cell.2015.06.043                           Cell  \n",
       "6  10.1016/j.cell.2015.06.043                           Cell  \n",
       "7  10.1016/j.cell.2015.06.043                           Cell  \n",
       "8  10.1016/j.cell.2015.06.043                           Cell  \n",
       "9  10.1016/j.cell.2015.06.043                           Cell  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/gregory/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n",
      "Error: list index out of range\n"
     ]
    }
   ],
   "source": [
    "# Save in project folder\n",
    "ppi_data.to_csv('../data/clean/biogrid_ppi_data.csv', index=False)\n",
    "gene_data.to_csv('../data/clean/gene_data.csv', index=False)\n",
    "\n",
    "# Save to Neo4j imports folder for LOAD CSV command\n",
    "ppi_data.to_csv(importDir + 'biogrid_ppi_data.csv', index=False)\n",
    "gene_data.to_csv(importDir + 'gene_data.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
